Split `--tunnel` flag to allow setting a tunneling protocol for a subset of traffic

pchaigno opened the pull request:
This pull request splits the `--tunnel` flag into two new flags to allow setting the tunneling protocol independently of the default routing mode (tunneling or native). This is useful for e.g. the egress gateway where users may want to use GENEVE instead of VXLAN but can't today.

Summary:
- Commit 1 introduces the two new flags.
- Commits 2--6 replace all uses of the `--tunnel` flag with the new flags.
- Commit 7 deprecates the `--tunnel` flag.
- Commit 8 updates the upgrade guide.

Test runs for the three issue_comment workflows modified in this PR:
- AWS-CNI: https://github.com/cilium/cilium/actions/runs/4543365761?pr=24561
- Datapath: https://github.com/cilium/cilium/actions/runs/4543365722?pr=24561
- Clustermesh: https://github.com/cilium/cilium/actions/runs/4543365755?pr=24561

```release-note
Deprecate `--tunnel` in favor of `--routing-mode` and `--tunnel-protocol`.
```

pchaigno committed:
helm: New flags for native routing and tunnel protocol

This commit introduces Helm flags nativeRouting and tunnelProtocol to
match routing-mode and tunnel-protocol. All use of the tunnel flag in
the Helm templates is also replaced by the new flags.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Remove all uses of tunnel Helm flag

All tests using the tunnel flag now use the new flags.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
docs: Replace all references to tunnel flags

This commit replaces all references to the agent and Helm flags in the
documentation by the new flags.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
daemon: Deprecate --tunnel

The option will be removed in v1.15.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
Introduce new flags for native routing and encap. protocol

The long-term plan is replace the --tunnel flag with --tunnel-protocol
and --routing-mode flags. In some cases (e.g., egress gateway
today), we need to setup an ad-hoc tunnel for some of the traffic
while in native routing mode. In such a configuration, it is not
possible today to pick the tunneling protocol and we default to VXLAN.
By splitting --tunnel, we will allow users to specify the tunneling
protocol even while running in native routing mode.

This commit simply introduces the two new flags. We fatal if
--routing-mode is set and --tunnel isn't empty.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
agent: Replace all use of --tunnel by the new flags

This commit replaces all use of --tunnel in the agent by the new flags
--routing-mode and --tunnel-protocol. For this to work, we translate
the --tunnel option, if not empty, into the appropriate equivalents
at startup. Thus, the following translations happen:

| Given options     | New options                          |
|-------------------|--------------------------------------|
| --tunnel=disabled | --routing-mode=native --tunnel=""    |
| --tunnel=vxlan    | --tunnel-protocol=vxlan --tunnel=""  |
| --tunnel=geneve   | --tunnel-protocol=geneve --tunnel="" |
| --tunnel=""       | --tunnel=""                          |

Options --routing-mode and --tunnel-protocol default to 'tunnel' and
'vxlan' respectively.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
docs: Document deprecation of tunnel and new flags

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
workflows: Remove all uses of tunnel flag

All tests using the tunnel flag now use the new flags.

Signed-off-by: Paul Chaignon <paul@cilium.io>

christarazi commented on Documentation/operations/upgrade.rst:
```diff
@@ -318,6 +318,22 @@ Annotations:
+Deprecated Options
+~~~~~~~~~~~~~~~~~~
+
+* The ``tunnel`` option was deprecated and will be removed in v1.14. To enable
```
Did you mean v1.15?

pchaigno replied:
Ouch, yes. That's because this is a rebase of a PR I already attempted to send into v1.13.

christarazi commented on pkg/option/config.go:
```diff
@@ -3082,8 +3112,14 @@ func (c *DaemonConfig) Populate(vp *viper.Viper) {
 	c.TunnelPort = vp.GetInt(TunnelPortName)
 
+	if c.Tunnel != "" && c.RoutingMode != defaults.RoutingMode {
+		log.Fatalf("Option --%s cannot be used in combination with --%s", RoutingMode, TunnelName)
```
Nit: Why fatal here when most* of the function is returning errors?

*: The other inconsistency is on L3310 and L3103. All others return errors.

pchaigno replied:
@christarazi Hm, I'm missing something here. There are 19 other fatals in this same function. The function also doesn't return anything anyway.

christarazi replied:
Oh I think I was looking at the wrong function, nevermind.

christarazi reviewed (approved):
Thanks for the very easy to read PR with nicely structured commits and a clear summary; a model example! :sparkles: 

A few minor non-blocking comments.

julianwiedmann commented on Documentation/operations/upgrade.rst:
```diff
@@ -341,6 +357,9 @@ Helm Options
 * The ``containerRuntime.integration`` value is being deprecated in favor of ``bpf.autoMount.enabled``.
+* Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` as been
+  deprecated in favor of ``routingMode`` and ``tunnelProtocol`` and will be
+  removed in v1.14.
```
```suggestion
* Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` has been
  deprecated in favor of ``routingMode`` and ``tunnelProtocol``, and will be
  removed in v1.14.
```

zacharysarah replied:
+1 on the typo fix, but no comma needed.

julianwiedmann commented on pkg/datapath/loader/base.go:
```diff
@@ -343,10 +343,10 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 		// Enable tunnel mode to vxlan if egress gateway is configured
 		// Tunnel is required for egress traffic under this config
-		args[initArgTunnelMode] = option.TunnelVXLAN
+		args[initArgTunnelMode] = option.Config.TunnelProtocol
```
The comment is no longer accurate (the protocol can now be either `vxlan` or `geneve`).

julianwiedmann commented on pkg/option/config.go:
```diff
@@ -2477,15 +2495,19 @@ func (c *DaemonConfig) AlwaysAllowLocalhost() bool {
+	// tunneling because, in unit tests, RoutingMode is usually not set and we
+	// would like for TunnelingEnabled to default to the actual default
+	// (tunneling is enabmled) in that case.
+	return c.RoutingMode != RoutingModeNative
```
I hope we can make this `s/TunnelingEnabled/RoutingModeIsTunnel` some day.

zacharysarah replied:
Agreed.

julianwiedmann reviewed (approved):
Thanks! Just a few small comments from browsing through. Could easily be addressed in a follow-on.

Re https://github.com/cilium/cilium/pull/24561/commits/55dc5914b39ef2f7ddf0b158b8f5d46c461151a5

> Options --routing-mode and --tunnel-protocol default to native and
vxlan respectively.

If I'm not mistaken, the `--routing-mode` actually defaults to `tunnel`?

tklauser commented on Documentation/operations/upgrade.rst:
```diff
@@ -318,6 +318,22 @@ Annotations:
+* The ``tunnel`` option was deprecated and will be removed in v1.14. To enable
+  native-routing mode, you can set ``routing-mode=native`` (previously
+  ``tunnel=disabled``). To configure the tunneling protocol, you can set
+  ``tunnel-protocol=geneve`` (previously ``tunnel=geneve``).
```
Should also mention `routing-mode=tunnel` here?
```suggestion
  ``tunnel=disabled``). To configure the tunneling protocol, you can set
  ``routing-mode=tunnel`` and ``tunnel-protocol=geneve`` (previously ``tunnel=geneve``).
```

pchaigno replied:
`routing-mode=tunnel` is already the default value (I said something different in a commit description; it was one of Julian's comments).

tklauser reviewed (approved):
Nice work on the clean commit history! Only one nit in addition to what Chris and Julian already pointed out.

Also, the commit message in the 3rd commit `helm: New flags for native routing and tunnel protocol` mentions a helm flag `nativeRouting.enabled` and a `enable-native-routing` flag. I guess these should say `routingMode` and `routing-mode` instead, respectively.

zacharysarah commented on Documentation/operations/upgrade.rst:
```diff
@@ -318,6 +318,22 @@ Annotations:
+Deprecated Options
+~~~~~~~~~~~~~~~~~~
+
+* The ``tunnel`` option was deprecated and will be removed in v1.14. To enable
```
I think it would help to specify the version in which the option was deprecated. For example:

```suggestion
* The ``tunnel`` option was deprecated in v1.1<x> and will be removed in v1.14. To enable
```

pchaigno replied:
I'm not sure I understand this one, but maybe the confusion comes from the typo here: this sentence should say "will be removed in v1.15", as per Chris's comment above.

Then, this whole section of the guide is about v1.14. Do we really want to specify that each item in the list is happening in v1.14?

pchaigno replied:
Hm, maybe the tense is the issue here. I keep using past tense when the other entries are using present or present continuous. Maybe that would help clarify we're still talking about v1.14?

zacharysarah replied:
> maybe the confusion comes from the typo here: this sentence should say "will be removed in v1.15", as per Chris's comment above.

There's no confusion about the version in which the option will be _removed_. I saw Chris's feedback before I reviewed. My suggestion was specifically to indicate the version in which the option was _deprecated_ (presumably v1.14).

> Then, this whole section of the guide is about v1.14. Do we really want to specify that each item in the list is happening in v1.14?

If the entire context of the document is explicitly specific to v1.14 and unlikely to be reused in a different, non-specific context, then I think it's fine to ignore my suggestion.

However: looking at the file, I see nothing to indicate a specific release in the title or description. Rather, I see that the page is implied to be version-specific because of the branch in which the file will reside when merged. 

The challenge, then, is that when this page is updated in 1.15 and subsequent versions, there will be nothing to indicate the version in which the options were removed--because the version in which deprecation occurred is implied, not explicit.

Does that make sense?

> I keep using past tense when the other entries are using present or present continuous. 

It's not a problem of verb tense agreement; it's a problem of implied vs. explicit. 

> Maybe that would help clarify we're still talking about v1.14?

I certainly think so. 😄

pchaigno replied:
This document isn't all specific to v1.14, but the section where those changes are in is for v1.14 only. See line 312 where the title is (this is a lot clearer on the rendered docs than on GitHub of course).

Once we branch for v1.15, a new v1.15 section will be added here. These entries will not be reused for other versions. This PR is also on purpose not marked for backport so it won't appear in older stable branches.

I'm not against being more specific on versions in sentences if we still think it's needed, but that doesn't look specific to this change. I'd prefer if we address it for all entries in a separate PR if we go that direction.

I'll fix up the bad verb tense.

joestringer replied:
(FWIW I agree with @zacharysarah here that we should just be more explicit, but yeah seems fine to just update them all at once in another PR to set that expectation moving forwards)

zacharysarah commented on Documentation/operations/upgrade.rst:
```diff
@@ -318,6 +318,22 @@ Annotations:
+
+* The ``tunnel`` option was deprecated and will be removed in v1.14. To enable
+  native-routing mode, you can set ``routing-mode=native`` (previously
+  ``tunnel=disabled``). To configure the tunneling protocol, you can set
```
Great example of "to accomplish X, do Y" writing. A small tweak will tighten it up.
```suggestion
  ``tunnel=disabled``). To configure the tunneling protocol, set
```

zacharysarah commented on Documentation/operations/upgrade.rst:
```diff
@@ -341,6 +357,9 @@ Helm Options
 * The ``securityContext`` for Hubble Relay now defaults to drop all
   capabilities and run as non-root user.
 * The ``containerRuntime.integration`` value is being deprecated in favor of ``bpf.autoMount.enabled``.
+* Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` as been
```
Again, I think it would help to specify the version in which the flag was deprecated. One typo fix:
```suggestion
* Following the deprecation of the ``tunnel`` agent flag in v1.1<x>, ``tunnel`` has been
```

zacharysarah commented on pkg/option/config.go:
```diff
@@ -2477,15 +2495,19 @@ func (c *DaemonConfig) AlwaysAllowLocalhost() bool {
+	// We check if routing mode is not native rather than checking if it's
+	// tunneling because, in unit tests, RoutingMode is usually not set and we
+	// would like for TunnelingEnabled to default to the actual default
+	// (tunneling is enabmled) in that case.
```
```suggestion
	// (tunneling is enabled) in that case.
```

zacharysarah reviewed (changes requested):
Nice work. ✨ Some suggestions and fixes, otherwise LGTM. 👍🏻

ldelossa reviewed (approved):
Looks good to me, and a welcome clarification of the flags.

I still see one reference to to `tunnel=disabled` in 

```
test/bigtcp/test.sh
34:    --set tunnel='disabled' \
```

Worth checking out?

tommyp1ckles commented on pkg/option/config.go:
```diff
@@ -2715,6 +2733,18 @@ func (c *DaemonConfig) Validate(vp *viper.Viper) error {
+	switch c.RoutingMode {
+	case RoutingModeNative, RoutingModeTunnel:
+	default:
+		return fmt.Errorf("invalid routing mode %q", c.RoutingMode)
```
minor: Should we include valid modes in the error message

tommyp1ckles reviewed (approved):
lg 👏

nbusseneau reviewed (approved):
Changes LGTM, are you also taking care of making changes accordingly to the Cilium CLI?

pchaigno committed:
test/egress: Set autoDirectNodeRoutes in native routing mode

When using VirtualBox, we need to use autoDirectNodeRoutes whenever we
are running Cilium in native routing mode. One case was missed in the
egress gateway tests.

Fixes: 4141a7e35bd ("test: add test for DSR with Geneve")
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
daemon, test: Require GENEVE tunneling if using GENEVE for DSR

We've recently enabled a new DSR mode where GENEVE is used to carry
information from the LB node to the backends. Let's enforce that GENEVE
be the tunneling mode as well for Cilium in that case, to avoid a weird
mix of VXLAN and GENEVE.

It remains possible to use native routing with GENEVE for DSR.

Mixing VXLAN and GENEVE should be possible, but until we hear of a use
case for that, I don't see a reason to support & maintain such a setup.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Cleanup overlay device if switching between protocols

We currently clean up the overlay devices if tunneling is disabled. We
however don't clean up them up if we are switching from VXLAN
(cilium_vxlan device) to GENEVE (cilium_geneve device). That can cause
issues where one device is already created and up and the other fails to
come up with "Address already in use".

This commit adds logic to properly clean up the old overlay device in
that case. The clean up logic also needs to be before the creation logic
as otherwise we will fail with the above error.

Signed-off-by: Paul Chaignon <paul@cilium.io>
